Feature/last steps for composer integration

teiling88 opened the pull request:
This pr is at the moment an informational pr for @following5 . 
With this changes I am testing this pr for oc.de: https://github.com/OpencachingDeutschland/oc-server3/pull/623

Please don't merge it yet.

following5 commented on okapi/services/logs/submit/WebService.php:
```diff
@@ -312,11 +312,9 @@ private static function _call(OkapiRequest $request)
                 $opt['html_purifier'] = Settings::get('OCDE_HTML_PURIFIER_SETTINGS');
-                require_once $GLOBALS['rootpath'] . 'lib2/OcHTMLPurifier.class.php';
 
                 $purifier = new \OcHTMLPurifier($opt);
```
**Fatal error:** Class 'OcHTMLPurifier' not found in /mnt/ocde/htdocs/okapi/services/logs/submit/WebService.php on line 319

Note: For testing methods with auth level > 2, you can add a debug username to the options; here in line 29, e.g.

`'min_auth_level' => 3, 'DEBUG_AS_USERNAME' => 'following5'`

teiling88 commented:
I can't reproduce this issue. 
Did you use this branch: https://github.com/OpencachingDeutschland/oc-server3/pull/623
After that do you make an composer install and updated the vendor/opencaching/okapi with this pr?

following5 commented:
> Did you use this branch: OpencachingDeutschland/oc-server3#623
> After that do you make an composer install and updated the vendor/opencaching/okapi with this pr?

Of course not. :-)

I will do another test according to your instructions.

Can you explain how the deployment from Okapi repository to OCDE code will work? 
* What's the role of https://github.com/OpencachingDeutschland/okapi-subtree-for-composer-usage?
* Can we set up some automatic deployment from Okapi repo to composer package?
* What will be the best way to locally test Okapi changes with OCDE code? Up to know, I create an Okapi-feature-branch in my local OCDE or OCPL repo where I test a new Okapi feature. Usually I have multiple of those OCDE- and/or OCPL-Okapi feature branches for testing different Okapi features under development. What would be the new development workflow with composer?

teiling88 commented:
The subtree repo is the source of opencaching/okapi and is represented here: https://packagist.org/packages/opencaching/okapi

> Can we set up some automatic deployment from Okapi repo to composer package?

already done - i created an automatic script for update the subtree repo.

> What will be the best way to locally test Okapi changes with OCDE code?

I develop in the okapi repo and created a deployment target to my oc.de development vm to vendor/opencaching/okapi. That works like a charm :-)

following5 commented:
> already done - i created an automatic script for update the subtree repo.

Is the script active? The subtree repo currently is two commits behind Okapi master.

> I develop in the okapi repo and created a deployment target to my oc.de development vm to vendor/opencaching/okapi. That works like a charm :-)

Would you explain to composer noobs how to do that? :-) Let's take this PR:

> Did you use this branch: OpencachingDeutschland/oc-server3#623
After that do you make an composer install and updated the vendor/opencaching/okapi with this pr?

First I have checked out the appropriate Git trees:
* `cd <OKAP-repo>`
* `git fetch teiling88`
* `git checkout teiling88/feature/last-steps-for-composer-integration`
* `cd <OCDE-repo>`
* `git fetch teiling88`
* `git checkout teiling88/feature/load-okapi-via-composer`

In the old workflow, for a quick test now I would do this:
* `ocopy` (which is a local script that mirrors the OKAPI code to the OCDE `htdocs/okapi` subtree)
* test
* `git reset --hard` to reset the OCDE code

(For an extended test, I would create an OCDE branch.)

All this is done directly on the host, where the repos reside; I do not need to touch the VM. Now how is the composer-based setup to test the Okapi feature branch?

Another typical bit of workflow is doing some small change in OCDE Okapi code for a quick test. E.g. I want to test `services/logs/submit`, which needs adding `'DEBUG_AS_DEVELOPER' => 'following5'` [here](https://github.com/opencaching/okapi/blob/master/okapi/services/logs/submit/WebService.php#L29).

Now I do this:
* add the `'DEBUG_AS_DEVELOPER' => 'following5'` in logs/submit/WebService.php
* test
* `git reset --hard`

What is the workflow with composer to do that? Essentially: What is the command to reset the code in `vendor/opencaching/okapi` to the current OCDE Git branch?

teiling88 commented:
> Is the script active? The subtree repo currently is two commits behind Okapi master.

No. The okapi repo has no tagged version so every commit would directly go live during an deployment on oc.de. 

> Would you explain to composer noobs how to do that? :-)
so we have the following directories:

- /sources/oc.de
- /sources/okapi

all changes from /sources/okapi/okapi are deployed with PHPStorm directly to sources/oc.de/htdocs/vendor/opencaching/okapi. 

To reset changes go to: htdocs/vendor/opencaching/okapi and use git reset --hard

The only change for you is the path where you can finde okapi :)

following5 commented:
> To reset changes go to: htdocs/vendor/opencaching/okapi and use git reset --hard
> The only change for you is the path where you can finde okapi :)

If that would work, it would be perfect. :-)

But Git does not know anything about files in the "vendor" tree. It will ignore them when checking out or resetting (of course - they are in `.gitignore`).

Is there some composer command which will reset `vendor/opencaching/okapi` after local modifications?

teiling88 commented:
> s there some composer command which will reset vendor/opencaching/okapi after local modifications?

Yes in vendor/opencaching/okpai should be a normal git repo where you can reset your changes with git reset --hard

following5 commented:
> Yes in vendor/opencaching/okpai should be a normal git repo where you can reset your changes with git reset --hard

Composer did only install the Okapi source files in vendor/opencaching/okapi. Should I manually clone the Okapi Git repo there?

If yes: How is that Okapi git repo synchronized to the OCDE code? If I do a Git reset for the Okapi repo, it must reset to the Okapi revision which corresponds to the currently active OCDE code version.

teiling88 commented:
> Composer did only install the Okapi source files in vendor/opencaching/okapi. Should I manually clone the Okapi Git repo there?

no - composer decides on composer.lock file which version should be to installed. If you want this for developing a checkout in this path should work.

Alternative a rm -rv vendor/ and a composer install should make a reset too for you.

following5 commented:
Here's why there is no Git repo in my vendor/opencaching/okapi path:

```
Loading composer repositories with package information
Installing dependencies (including require-dev) from lock file
  - Installing opencaching/okapi (dev-master c1eeaab)
    Cloning c1eeaab91c3c84fe5fd64d98176fbac76cf6ab2e
    c1eeaab91c3c84fe5fd64d98176fbac76cf6ab2e is gone (history was rewritten?)
    Failed to download opencaching/okapi from source: Failed to execute git checkout 'c1eeaab91c3c84fe5fd64d98176fbac76cf6ab2e' -- && git reset --hard 'c1eeaab91c3c84fe5fd64d98176fbac76cf6ab2e' --

fatal: reference is not a tree: c1eeaab91c3c84fe5fd64d98176fbac76cf6ab2e

    Now trying to download from dist
  - Installing opencaching/okapi (dev-master c1eeaab)
    Loading from cache
```

Any ideas what goes wrong?

Now after doing this ...
> use this branch: OpencachingDeutschland/oc-server3#623
After that do you make an composer install and updated the vendor/opencaching/okapi with this pr

I get this Okapi error:

fopen(/mnt/ocde/htdocs/vendor/opencaching/okapi/../var/okapi/okapi-lock-cronjobs-pre-request)

The correct path would be /mnt/ocde/htdocs/vendor/opencaching/okapi/../../../var/okapi/okapi-lock-cronjobs-pre-request.

teiling88 commented:
> Any ideas what goes wrong?

i messed up the subtree repo

teiling88 commented:
So i corrected the subtree repo and updated https://github.com/OpencachingDeutschland/oc-server3/pull/623

> ErrorException:
> fopen(/mnt/ocde/htdocs/vendor/opencaching/okapi/../var/okapi/okapi-lock-cronjobs-pre-request)

How can I reproduce this?

following5 commented:
> So i corrected the subtree repo and updated OpencachingDeutschland/oc-server3#623

I have fetched that and checked it out, deleted the complete vendor tree and did an `composer install`. Again there is no Git repo in vendor/opencaching/okapi, just the Okapi files. Then I deleted vendor/opencaching, did `composer clear-cache` and `composer install` and again - no Okapi Git repo, just the Okapi files.

This is what composer said when installing:

```
  - Installing opencaching/okapi (1.1.0)
    Downloading: 100%
```
With the old version it said "Cloning" (see above).

> How can I reproduce this?

It appeared with any OKAPI request. The Okapi `VAR_DIR` setting was wrong. It is initialized in settings-dist.inc.php as `$opt['rootpath'] . 'var/okapi'`. Do you have some special settings for that on your system? I did'n see any in the Vagrant [settings template](https://github.com/OpencachingDeutschland/oc-server3/blob/development/dev-ops/local.team-opencaching.de/templates/config2-settings.inc.php.tpl).

I have now added this to my config2/settings.inc.php:

`$opt['okapi']['var_dir'] = $opt['rootpath'] . '../../var/okapi';`

Now the error message is gone. Also, the HTMLpurifier call in services/caches/submit now runs withour error.

I will do some more tests, especially with OCPL code.

following5 commented:
> `$opt['okapi']['var_dir'] = $opt['rootpath'] . '../../var/okapi';`

This doesn't look right. ` $opt['rootpath']` should point to htdocs, but it points to htdocs/vendor/opencaching.

teiling88 commented:
I think there is no need for a global variable rootpath or?

following5 commented:
Can you fix the issue with composer installation of the Okapi Git repo in OCDE vendor/opencaching/okapi? Then I would go for merge here.

teiling88 commented:
Can you please recheck this? I hope i changed it correctly. 

Thank you!

following5 commented:
* composer clear-cache
* rm -rf vendor/opencaching
* composer install

```
Loading composer repositories with package information
Installing dependencies (including require-dev) from lock file
  - Installing opencaching/okapi (1.1.0)
    Downloading: 100%

Generating autoload files
```

=> No .git repository in vendor/opencaching/okapi.

teiling88 commented:
> No .git repository in vendor/opencaching/okapi. :-(
my fault - it was deployed from phpstorm - sorry.

following5 commented:
It is still the same. Could you please test this on your system?

* cd htdocs
* composer clear-cache
* rm -rf vendor/opencaching
* composer install
* ls vendor/opencaching/okapi/.git

teiling88 commented:
> It is still the same. Could you please test this on your system?

that was I said - the .git repo was deployed from phpstorm. With a composer install there is no .git directory. You are right.

following5 commented:
So for Okapi developers who don't use phpStorm the Git capability for OCDE testing is lost when migrating to composer? This would be a big drawback.

Is it possible to include the Okapi .git repo into the composer package?

Some time ago you suggested to integrate Okapi via "Git subcommits" into the OCDE Git repo. Would that be an alternative to deployment via composer?

following5 commented:
> Is it possible to include the Okapi .git repo into the composer package?

When cloned via Git, the Okapi repo is just a 3 MB download. So in terms of size it should not matter.

teiling88 commented:
> So for Okapi developers who don't use phpStorm the Git capability for OCDE testing is lost when migrating to composer? This would be a big drawback.

No it should be very easy four you to symlink this or create with other IDE's a remote deployment project. I see no drawback there. It is only a little change in the development process for somebody who didn't use composer or composer packages so far.

nlubisch commented:
I totally agree with @teiling88. This is our daily and common workflow not only for OC.

mirsch commented:
I did not read the whole story, but if I get it right the problem is you want to commit inside the vendor dir right?
try ```composer update opencaching/okapi --prefer-source``` this should give you the repo inside vendor.
If you want to use your own fork for development you may add your repository to ```composer.json``` to override the default. E.g.
```
...
"repositories": [
        {
            "type": "git",
            "url": "git@github.com:following5/okapi.git"
        }
]
...
```

and select the correct branch also in ```composer.json``` e.g.
```
...
"require": {
        "opencaching/okapi": "dev-YOUR-BRANCH-NAME"
}
...
```
remember to ```composer update opencaching/okapi``` after modifying ```composer.json```

following5 commented:
@teiling88 @nlubisch 
Please undestand tha Okapi is a project of it's own and has it's own development team, culture and requirements, which in many aspects differ from what you are used to. As Okapi and OCDE / OCPL depend on each other, we always need to cooperate and find a solution which fits the requirements of all - Okapi, OCDE and OCPL.

I am doing my best in helping to adjust Okapi to the new requirements of the new OCDE development team, but I need to make sure that Okapi development is not disrupted by this change.

@mirsch 
> try composer update opencaching/okapi --prefer-source this should give you the repo inside vendor.

That took about 10 minutes and ended with:

> PHP Fatal error:  Allowed memory size of 1073741824 bytes exhausted (tried to allocate 72 bytes) in phar:///usr/bin/composer/src/Composer/DependencyResolver/Rule.php on line 60

@ all
Why is it so important to deploy Okapi through composer at OCDE? Which problems arise from replicating Okapi code into the OCDE Git repo?

Note that Okapi runs only as part of OCDE or OCPL code, it is not a general-purpose library like all the other libraries that are managed by composer at OCDE. Okapi and OCDE/OCPL code are tied closely to each other, as they share the database and also some code. So it looks not that far-fetched to me to keep al the code in one repo.

following5 commented:
> No it should be very easy four you to symlink this

The Okapi Git versions must be tied to OCDE Git versions: If another OCDE version is checked out, then the Okapi Git must point to the corresponding Okapi commit. This will not work with a symlink.

following5 commented:
After some more thoughts:

There are two fundamentally different ways to include Okapi code into OCDE code tree using Git.

1. The current way is to treat OCDE+Okapi as one "combined project", by keeping all the code in one repo. This allows to easily handle corresponding OCDE+Okapi code versions: Checkout an OCDE version, and there is the matching Okapi version.

2. The other way - as proposed by teiling88 - is to keep two separate Git projects in the OCDE code tree. This has the advantage that you can commit changes made in vendor/opencaching/okapi directly into the Okapi repo. Also, if OCPL would adopt the OCDE solution for Okapi deployment, it would allow to share Okapi feature branches between the developers' OCDE and and OCPL installations.

(So my question "Which problems arise from replicating Okapi code into the OCDE Git repo?" is answered: It forces developers to locally replicate code between multiple repositories.)

I think the best solution would be to 
* remove Okapi code from OCDE repo, as proposed by teiling88,
* somehow include the Okapi git repository in the local OCDE code tree (e.g. by symlink or IDE remote deployment, as proposed by teiling88),
* somehow tie the active Okapi code version to active OCDE code version in an _IDE-independent_ way.

So when checking out an OCDE code version, this must trigger a checkout of the corresponding Okapi version. If the Okapi code subtree is dirty and the changes would conflict with the checkout, the OCDE checkout must be blocked in the same way as if the OCDE code were dirty.

How to do this technically?

following5 commented:
(Note to the email-repliers: I have edited my previous post and added some important words to "1.":  _when developing_ and _with all Git features available for the "combined project"_.)

mirsch commented:
Im on the go and can not go deeper into this but maybe "git submodule" is what you are looking for.

mirsch commented:
Never the less composer is the right tool to handle such dependencies. It's exactly what it was build for.

following5 commented:
@teiling88 
> in vendor/opencaching/okpai should be a normal git repo

> the .git repo was deployed from phpstorm.

Are you sure that this .git repo is in vendor/opencaching/okapi, and that it works? vendor/opencaching/okapi just contains the contents of the 'okapi' subdirectory of the Okapi code tree. So if the Okapi .git dir is replicated (or symlinked) to vendor/opencaching/okapi, it does not work.

mirsch commented:
@teiling88 why is composer.json not placed in the root of okapi repo? I think the subtree split is not necessary and it avoids development of okapi inside the project because you dont get the complete repo.

following5 commented:
> Im on the go and can not go deeper into this but maybe "git submodule" is what you are looking for.

As far as I understand the Git submodules, this would need to be setup in the OCDE Git repo and replace the composer integration.

But I think there is a way to tie Okapi Git versions to OCDE Git versions using composer: [Here](https://github.com/OpencachingDeutschland/oc-server3/pull/623/commits/2d5e4a2830129cb0c680b49621724d68cca8b8c2), the original [Okapi version number](https://github.com/OpencachingDeutschland/oc-server3/blob/development/htdocs/okapi/core/Okapi.php#L32) must be used. (Why a proprietary numbering? Okapi version numbers are integers). From this version number, the Okapi Git revision (commit ID) can be calculated and checked out after `composer install`.

Or even better: Directly include the Okapi commit ID into the composer version numbers, e.g. `1609-60c156c`. This would exactly reflect Okapis own versioning scheme, see [here](https://www.opencaching.de/okapi/introduction.html) at the upper left. Both values need to be calculated anyway for creating the composer package, because they must be inserted in core/Okapi.php.

following5 commented:
Oh well ... when the Git commit ID is already in core/Okapi.php, then we already have everything we need to check it out after `composer install`. (But I still would prefer to use the original Okapi version numbers for composer! "1.0.1" has not any meaning for Okapi.

BUT **where are the Okapi version numbers in core/Okapi.php**? After checkout by composer, I have `null` values there.

following5 commented:
> Oh well ... when the Git commit ID is already in core/Okapi.php, then we already have everything we need to check it out after composer install

... but it gets lost after direct checkout out of another Okapi version. So the original Okapi version number (best with appended commit ID) should be in htdocs/composer.json.

teiling88 commented:
> Directly include the Okapi commit ID into the composer version numbers, e.g. 1609-60c156c.

please have a deeper look into composer and composer.lock. There should be the connection between oc.de and okapi git version. So after every git checkout you just need a composer install for that. 

After my vacation in the end of the next week - I will provide more Information.

following5 commented:
> please have a deeper look into composer and composer.lock. There should be the connection between oc.de and okapi git version. So after every git checkout you just need a composer install for that.

I am talking about the active version in vendor/opencaching/okapi/.git

following5 commented:
> please have a deeper look into composer and composer.lock. There should be the connection between oc.de and okapi git version. So after every git checkout you just need a composer install for that.

I am talking about the version in vendor/opencaching/okapi/.git.

So if you do 
* composer install
* cd vendor/opencaching/okapi/
* git rev-parse HEAD

it will show the commit ID of the Okapi version which was installed by composer?

following5 commented:
To sum up, there currently are these issues:

* The original Okapi version numbers should be used for composer.
* composer installs only a subtree of Okapi repo, so that is it not possible to use Git in opencaching/vendor/okapi.
* Okapi version number and commit ID is missing in core/Okapi.php when installed by composer

Have a nice vacation! :-)

teiling88 commented:
> To sum up, there currently are these issues:

I think this issues are not relevant for this pr. So please feel free to merge it when possible.

Your mentioned issues should be discussed on this pr: https://github.com/OpencachingDeutschland/oc-server3/pull/623

following5 commented:
The second issue affects this PR, because it will change the path where the Okapi files reside (vendor/opencaching/okapi => vendor/opencaching/okapi/okapi).

I would prefer to keep the discussion about the whole composer thing at one place and in Okapi project, because it affects Okapi developers more than OCDE developers: All Okapi developers need to develop and test on the OCDE platform, but not all OCDE developers need to do Okapi development.

My concern here is that - if I merge this - the main issue (tying the vendor/opencaching/okapi/okapi/.git version to the OCDE code version) may be ignored, which would break Okapi development workflow.

If you change the Okapi composer versioning scheme so that it includes Okapi versioning (e.g. "1646-bd14610" for the current version), I would be fine with proceeding here and merging this PR.

teiling88 commented:
The mentioned composer versions will be removed. It was some kind of testing for packagist.org. So we will reference to dev-master and the used commit id can be found in composer.lock so every okapi update is saved in composer.lock. So there is a direct connection between oc.de and okapi.

following5 commented:
Okay!

And can you confirm that https://github.com/OpencachingDeutschland/oc-server3/pull/623 will not be merged before the other two issues are resolved? So when the OCDE PR is merged, we can continue to use Git for Okapi code in vendor/opencaching/okapi?

teiling88 commented:
We had today a oc.de developer meeting. We discussed a less complex solution. I will extend this pr for this. So we will need no subtree repo any more

teiling88 commented:
Please have a new look into https://github.com/OpencachingDeutschland/oc-server3/pull/623

following5 commented:
Works great. Thanks!
